### PR TITLE
Add JAX support (stormlog.jax)

### LINKED ISSUE #124: Add JAX support under a dedicated stormlog.jax package
## Summary
Add JAX support as a dedicated `stormlog.jax` package instead of forcing it through the PyTorch tracker surface.

## Why
JAX has viable memory/profiling primitives, but it is a different integration shape from PyTorch:
- `jax.Device.memory_stats()` exposes device memory statistics keyed by name.
- `jax.profiler.save_device_memory_profile()` can emit a device memory profile artifact.

Relevant docs:
- https://docs.jax.dev/en/latest/_autosummary/jax.Device.html
- https://docs.jax.dev/en/latest/_autosummary/jax.profiler.save_device_memory_profile.html

Stormlog’s current architecture already separates framework-specific runtime logic into subpackages (`stormlog` for PyTorch and `stormlog.tensorflow` for TensorFlow). JAX should follow the same pattern rather than being forced through CUDA-oriented PyTorch assumptions.

## Proposed direction
Create a dedicated `stormlog.jax` surface that reuses the shared session/telemetry/artifact layer while owning its own runtime-facing tracker/profiler code.

First milestone:
- `stormlog.jax.MemoryTracker`
- basic JAX CLI entrypoint or subcommands
- session-aware telemetry export
- device-memory profile artifact export where supported

Later follow-ups:
- JAX-specific analysis helpers and recommendations
- distributed JAX coverage
- JAX cookbook/testing guidance

## Scope
- New framework-specific package boundary for JAX.
- Tracker support backed by documented JAX runtime APIs.
- Shared session/telemetry compatibility with existing CLI/TUI loading.
- Explicit capability reporting where allocator-native CUDA analysis does not apply.

## Non-goals
- Forcing JAX support into the existing PyTorch collector/tracker path.
- Claiming feature parity with CUDA-native attribution/history on day one.
- Mixing JAX-specific runtime decisions into `stormlog.tensorflow`.

## Acceptance criteria
- JAX workloads can emit Stormlog-compatible session-aware telemetry.
- JAX memory/profile artifacts can be produced with documented upstream APIs.
- Shared loading/analysis paths consume the new artifacts where the schema fits.
- The implementation lands behind a clear `stormlog.jax` package boundary.

### PR DESCRIPTION
**Pull request by DeeNA95**
## Summary

Dedicated `stormlog.jax` package for JAX memory profiling/tracking, following the same subpackage pattern as `stormlog` (PyTorch) and `stormlog.tensorflow` (TensorFlow). Closes #124.

### Package surface (`stormlog.jax.*`)

| Module | Exports |
|--------|---------|
| `tracker.py` | `MemoryTracker`, `MemoryWatchdog`, `TrackingResult` |
| `profiler.py` | `JAXMemoryProfiler`, `MemorySnapshot`, `ProfileResult`, global profiler singleton |
| `analyzer.py` | `MemoryAnalyzer` (full parity: leak detection, fragmentation, efficiency scoring, gap analysis, collective attribution) |
| `context_profiler.py` | `JAXProfiler`, `ProfiledFunction`, `profile_function`, `profile_context` |
| `diagnose.py` | `run_diagnose()` — portable diagnostic bundle builder |
| `utils.py` | `get_device_info`, `get_system_info`, `format_memory`, `validate_jax_environment` |
| `visualizer.py` | `MemoryVisualizer` (matplotlib + plotly) |
| `pprof_parser.py` | `parse_jax_memory_profile()` — parses gzipped pprof protobuf |
| `attributed_viz.py` | `render_jax_attributed_html()` — self-contained Graphviz HTML dashboard |

### CLI (`jaxmemprof`)

```bash
jaxmemprof info        # system + device info
jaxmemprof track       # background tracking with telemetry/OOM/weave
jaxmemprof monitor     # real-time memory display
jaxmemprof diagnose    # portable diagnostic bundle
jaxmemprof analyze     # analyze saved tracking JSON
```

### Key implementation details

- Uses `jax.Device.memory_stats()` with mandatory XLA sync (`block_until_ready()`) before every read
- `save_device_memory_profile()` calls `jax.profiler.save_device_memory_profile()` with runtime capability guards
- Reserved memory explicitly aliased to allocated in telemetry (JAX XLA has no separate reserved pool), with docstring warnings for downstream consumers
- OOM flight recorder enriches dump manifests with `.prof` artifacts and auto-generates interactive HTML call-graph dashboards
- Phase tracking (`enter_phase` / `phase` context manager) integrated via shared `stormlog.phases` infrastructure

### Acceptance criteria (from #124)

- [x] JAX workloads can emit Stormlog-compatible session-aware telemetry
- [x] JAX memory/profile artifacts can be produced with documented upstream APIs
- [x] Shared loading/analysis paths consume the new artifacts where the schema fits
- [x] The implementation lands behind a clear `stormlog.jax` package boundary

### Non-goals verified
- [x] No JAX support forced into the PyTorch `stormlog` tracker path
- [x] No claim of CUDA-native attribution/history parity
- [x] No JAX-specific runtime decisions mixed into `stormlog.tensorflow`

### Deferred (future PRs)
- Distributed JAX coverage (infrastructure in place, multi-device scenarios untested)
- JAX cookbook / testing guidance

### CODE REVIEW

File: stormlog/jax/tracker.py [hunk 1]
```diff
@@ -0,0 +1,1334 @@
...
+            # sync that can slightly perturb workload timing at high sample
+            # rates.  At the default 1 s interval the overhead should be negligible.
+            if self._sync_sentinel is not None:
+                self._sync_sentinel.block_until_ready()
+            elif hasattr(jax.numpy, "zeros"):
+                jax.numpy.zeros((), device=self._device).block_until_ready()
+            stats = self._device.memory_stats()
+            if stats:
+                return int(stats.get("bytes_in_use", 0))
+        except Exception:
```
Comment (Silas-Asamoah) [inline stormlog/jax/tracker.py]: Issue: this swallows every failure from the sync or `memory_stats()` read and returns `0`, so `_run_tracking_iteration()` records a healthy zero-memory sample instead of moving collector health into a degraded/failed state. That can make broken JAX telemetry look like a real memory drop. We could let these exceptions propagate into the existing collector-health transition path, or return an explicit failed sample state and avoid emitting synthetic zero usage.

File: stormlog/jax/tracker.py [hunk 2]
```diff
@@ -0,0 +1,1334 @@
...
+        self,
+        *,
+        timestamp: float,
+        memory_bytes: int,
+        event_type: str = "sample",
+        context: Optional[str] = None,
+        metadata: Optional[Dict[str, Any]] = None,
+    ) -> Dict[str, Any]:
+        sampling_interval_ms = int(round(self.sampling_interval * 1000))
+        # JAX's XLA runtime does not expose a separate "reserved but
```
Comment (Silas-Asamoah) [inline stormlog/jax/tracker.py]: Issue: setting `allocator_reserved_bytes` equal to `allocator_allocated_bytes` makes the shared allocator counters look exact while hiding JAX/XLA reserved or preallocated memory. JAX documents GPU memory preallocation and allocation controls here: https://docs.jax.dev/en/latest/gpu_memory_allocation.html, and the `Device` API exposes `memory_stats()` here: https://docs.jax.dev/en/latest/_autosummary/jax.Device.html. Since this PR already reads `bytes_reserved` in the profiler path, the tracker/diagnose projection should use that field when present, and only fall back to aliasing with metadata that marks the value as approximate.

File: stormlog/jax/cli.py [hunk 3]
```diff
@@ -0,0 +1,787 @@
...
+    print(f"Python Version: {system_info['python_version']}")
+    print(f"CPU Count: {system_info['cpu_count']}")
+
+    if "total_memory_gb" in system_info:
+        print(f"Total System Memory: {system_info['total_memory_gb']:.2f} GB")
+        if "available_memory_gb" in system_info:
+            print(f"Available Memory: {system_info['available_memory_gb']:.2f} GB")
+
+    backend_info = system_info.get("backend", {})
+    device_count = backend_info.get("runtime_gpu_count", 0)
```
Comment (Silas-Asamoah) [inline stormlog/jax/cli.py]: Issue: `get_backend_info()` emits `device_count`, but this command reads `runtime_gpu_count`, so `jaxmemprof info` will report zero available devices when fed the real helper output. The same schema mismatch appears below with `device_kind` versus `kind`. We should align the CLI with `stormlog.jax.utils` and update the test mocks to use the real helper contract.

File: stormlog/jax/cli.py [hunk 4]
```diff
@@ -0,0 +1,787 @@
...
+        print("Error: --interval must be > 0", file=sys.stderr)
+        return 1
+
+    wandb_config = _resolve_wandb_config(args)
+    if wandb_config is None:
+        return 1
+
+    command_line = " ".join(sys.argv)
+    try:
+        artifact_dir, exit_code = run_diagnose(
```
Comment (Silas-Asamoah) [inline stormlog/jax/cli.py:446]: Issue: `run_diagnose` is imported only when JAX is available, but this command does not have the `JAX_AVAILABLE` guard used by `monitor` and `track`. A base install that exposes `jaxmemprof diagnose` without the JAX extra can hit an unbound name instead of the clean error path. This should mirror the other commands and add a no-JAX CLI test for diagnose.

**Review by Silas-Asamoah — COMMENTED** [review]
The JAX support boundary is generally in the right shape. I left a few inline notes where the JAX runtime counters and CLI behavior do not yet line up cleanly with the shared Stormlog contract.

File: stormlog/jax/tracker.py (hunk 1 above)
Comment (DeeNA95) [reply stormlog/jax/tracker.py]: Removed the try/except: pass from `_get_current_device_memory` entirely, exceptions now bubble up to `_run_tracking_iteration` which already has the `_transition_to_failure` path wired up. No more silent zero-memory samples when the sync or `memory_stats()` read breaks.

File: stormlog/jax/tracker.py (hunk 2 above)
Comment (DeeNA95) [reply stormlog/jax/tracker.py]: `_get_current_device_memory` now captures `bytes_reserved` from `memory_stats()` when present and stores it on the instance. `_build_telemetry_event_record` uses that real value for `allocator_reserved_bytes`. When `bytes_reserved` isn't available (e.g. CPU fallback), we still alias to allocated but stamp the metadata with `"allocator_reserved_approximate": True` so downstream consumers can distinguish.

File: stormlog/jax/cli.py (hunk 3 above)
Comment (DeeNA95) [reply stormlog/jax/cli.py]: Updated both lookups and the corresponding test mocks.

File: stormlog/jax/cli.py (hunk 4 above)
Comment (DeeNA95) [reply stormlog/jax/cli.py:446]: Added the same if not `JAX_AVAILABLE `guard that `cmd_monitor` and `cmd_track use`. Also added `test_cmd_diagnose_jax_not_available` to cover the no-JAX path.

File: stormlog/jax/diagnose.py [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+    device_info = get_device_info(device_index)
+    backend_info = get_backend_info()
+    backend = backend_info.get("runtime_backend", "cpu")
+    stats = device_info.get("memory_stats", {})
+
+    allocated = int(stats.get("bytes_in_use", 0) or 0)
+    peak = int(stats.get("peak_bytes_in_use", 0) or 0)
+    limit_bytes = int(stats.get("bytes_limit", 0) or 0)
+
+    # JAX has no separate reserved pool; alias to allocated (like TF)
```
Comment (Silas-Asamoah) [inline stormlog/jax/diagnose.py]: Issue: the tracker path now uses JAX `bytes_reserved`, but `jaxmemprof diagnose` still aliases reserved memory to allocated memory here and later reports `reserved_bytes` from `allocated`. That keeps diagnose bundles showing zero allocator gap even when JAX/XLA has reserved or preallocated memory. JAX documents GPU memory preallocation and allocation controls here: https://docs.jax.dev/en/latest/gpu_memory_allocation.html, and the `Device.memory_stats()` API is documented here: https://docs.jax.dev/en/latest/_autosummary/jax.Device.html. We should read `bytes_reserved` from `memory_stats()` when present, and only fall back to aliasing with an explicit approximation marker.

File: stormlog/jax/tracker.py [hunk 6]
```diff
@@ -0,0 +1,1347 @@
...
+            "world_size": self.distributed_identity["world_size"],
+        }
+        if metadata:
+            dump_metadata.update(metadata)
+
+        current_memory = self._status_memory_value()
+        dump_metadata.update(
+            {
+                "sample_allocated_bytes": current_memory,
+                "sample_reserved_bytes": current_memory,
```
Comment (Silas-Asamoah) [inline stormlog/jax/tracker.py]: Issue: OOM dump metadata still records `sample_reserved_bytes` from `current_memory`, so OOM bundles can disagree with the fixed telemetry samples whenever `bytes_reserved` differs from `bytes_in_use`. We could reuse the last sampled reserved value here, falling back to `current_memory` only when reserved memory is unavailable and marking that fallback as approximate.

**Review by Silas-Asamoah — COMMENTED** [review]
The previous fixes handled collector health and CLI behavior. I left two follow-up notes where reserved-memory handling is still inconsistent with the updated tracker projection.

**Comment by DeeNA95** [comment]
### Updates

 **JAX Diagnostics (`stormlog/jax/diagnose.py`)**
    * Updated `build_diagnostic_summary()` to retrieve JAX's actual `bytes_reserved` metric, falling back to
  allocated memory with the `"allocator_reserved_approximate": True` metadata flag only if unavailable.
    * Updated `run_timeline_capture()` to reconstruct the timeline by querying `allocator_reserved_bytes` from
  individual telemetry events instead of duplicating the allocated timeline.

 **OOM Flight Recorder (`stormlog/jax/tracker.py`)**
    * Fixed `handle_exception()` to utilize the tracker's actual `_last_reserved_bytes` for the OOM dump
  `sample_reserved_bytes` metadata, setting `"allocator_reserved_approximate": True` as a fallback.

 **CLI Information (`stormlog/jax/cli.py`)**
    * Updated the `jaxmemprof info` command to display the device's `Reserved Memory` alongside other metrics when
  available.

**Review by Silas-Asamoah — APPROVED** [review]
